lm_eval/utils.py

```diff
@@ -84,13 +85,19 @@ def complete_code(
         with torch.no_grad():
             if task.stop_words:
                 gen_kwargs["stopping_criteria"][0].start_length = batch["ids"].shape[-1]
+
+            if batch["ids"].shape[0]==1:
+                batch["ids"] = batch["ids"][:,:batch["input_len"]]
+                batch["attention_mask"] = batch["attention_mask"][:,:batch["input_len"]]
+            
             generated_tokens = accelerator.unwrap_model(model).generate(
-                input_ids=batch["ids"][:, : batch["input_len"]],
-                num_return_sequences=batch_size,
+                input_ids=batch["ids"],
+                attention_mask=batch["attention_mask"],
+                num_return_sequences=num_return_sequences,
                 **gen_kwargs,
             )
-            # each task is generated batch_size times
-            generated_tasks = batch["task_id"].repeat(batch_size)
+            # each task is generated num_return_sequences times
+            generated_tasks = batch["task_id"].repeat(num_return_sequences)
```

Comment:
In case `batch_size!=1` for example 2 and task_ids is [0,1], `repeat` gives [0,1,0,1] but we need [0,0,1,1] since that's the actual order of generations, `repeat_interleave` does this.
```suggestion
            generated_tasks = batch["task_id"].repeat_interleave(num_return_sequences)
```
Without this the scores are very low when `n_samples=num_return_sequences` and `batch_size>1` which is when one GPU gets different tasks:
```bash
accelerate launch  main.py    --model bigcode/santacoder    --tasks humaneval   --max_length_generation 256\
    --limit 20   --temperature 0.2    --do_sample True    --n_samples 20    --batch_size 1   --num_return_sequences 20    --allow_code_execution  True
```
gives `pass@1=0.265`.

But
```bash
accelerate launch  main.py    --model bigcode/santacoder    --tasks humaneval   --max_length_generation 256 \
   --limit 20   --temperature 0.2    --do_sample True    --n_samples 20    --batch_size 2   --num_return_sequences 20    --allow_code_execution  True
```
gives `pass@1=0.129`.

After the fix `pass@1` becomes 0.2575 and if we use `batch_size=4` it is 0.227, so there might still be something to fix.